iroh/src/downloader.rs

```diff
@@ -956,19 +1104,58 @@ impl<'a> Iterator for ProviderIter<'a> {
 
 impl ProviderMap {
     /// Get candidates to download this hash.
-    fn get_candidates(&self, hash: &Hash) -> impl Iterator<Item = &PublicKey> {
-        let inner = self.candidates.get(hash).map(|peer_set| peer_set.iter());
+    fn get_candidates(&self, hash: &Hash) -> impl Iterator<Item = &PeerInfo> {
+        let inner = self.candidates.get(hash).map(|peers| peers.values());
         ProviderIter { inner }
     }
 
     /// Register peers for a hash. Should only be done for hashes we care to download.
-    fn add_peers(&mut self, hash: Hash, peers: &[PublicKey]) {
-        self.candidates.entry(hash).or_default().extend(peers)
+    fn add_peers(&mut self, hash: Hash, peers: &[PeerInfo]) {
+        let entry = self.candidates.entry(hash).or_default();
+        for peer in peers {
+            entry
+                .entry(peer.node_id)
+                .and_modify(|existing| existing.role = (existing.role).max(peer.role))
+                .or_insert(*peer);
+            if let PeerRole::Provider = peer.role {
+                self.provider_hashes_by_peer
+                    .entry(peer.node_id)
+                    .or_default()
+                    .push_back(hash);
+            }
+        }
+    }
+
+    /// Get the next provider hash for a peer.
+    ///
+    /// I.e. get the next hash that was added with [`PeerRole::Provider`] for this peer.
+    fn get_next_provider_hash_for_peer(&mut self, peer: &PublicKey) -> Option<Hash> {
+        self.provider_hashes_by_peer
+            .get(peer)
+            .and_then(|hashes| hashes.front())
+            .copied()
```

Comment:
you can save the call on ended download to `move_hash_to_back` by doing it here. Also you can use `rotate_left` (or right, not sure which one)